lib/Dialect/FIRRTL/Transforms/CheckCombLoops.cpp

```diff
@@ -25,533 +26,472 @@
-        }
-      currentPath.pop_back();
-    }
-    currentPath.pop_back();
+      while (!dfsStack.empty()) {
+        auto dfsVal = dfsStack.back();
+        if (!visiting.contains(dfsVal)) {
+          unsigned dfsSize = dfsStack.size();
 
-    visiting = std::move(tmpVisiting);
-    // Finished discovering all the reachable nodes from node.
-    visited.insert(node);
-    return success();
-  }
+          LLVM_DEBUG(llvm::dbgs() << "\n Stack pop :"
+                                  << getFieldName(Node(dfsVal, 0)).first << ","
+                                  << dfsVal;);
 
-  /// Customized implementation of getFieldRefFromValue. If `field` is a
-  /// subfield, then this will return the base aggregate Value and the
-  /// corresponding fieldId. This will return nullptr if thie field is a result
-  /// of a register and can be ignored.
-  FieldRef getFieldRef(Value field) {
-    if (!isa_and_nonnull<SubfieldOp, SubaccessOp, SubindexOp>(
-            field.getDefiningOp()))
-      return FieldRef(field, 0);
-    auto iter = leafToBaseValMap.find(field);
-    if (iter == leafToBaseValMap.end())
-      return FieldRef(nullptr, 0);
-    auto baseVal = iter->second;
-    auto storIter = valToStorageLocIdMap.find(baseVal);
-    if (storIter == valToStorageLocIdMap.end())
-      return FieldRef(nullptr, 0);
-    auto baseStorId = storIter->getSecond().front();
-    auto fieldStorIter = valToStorageLocIdMap.find(field);
-    if (fieldStorIter == valToStorageLocIdMap.end())
-      return FieldRef(nullptr, 0);
-    auto fieldStorId = fieldStorIter->getSecond().front();
-
-    return FieldRef(baseVal, fieldStorId - baseStorId);
-  }
+          // Visiting set will contain all the values which alias with the
+          // dfsVal, this is required to detect back edges to aliasing Values.
+          // That is fieldRefs that can refer to the same memory location.
+          visiting.appendEmpty();
+          // Record all the children of `dfsVal`.
+          SmallVector<Value, 8> children;
+          // If this is an input port field, then record it. This is used to
+          // discover paths from input to output ports. Only the last input port
+          // that is visited on the DFS traversal is recorded.
+          SmallVector<Node, 2> inputArgFieldsTemp;
+          // All the Values that refer to the same FieldRef are added to the
+          // aliasingValues.
+          SmallVector<Value> aliasingValues = {dfsVal};
+          // If `dfsVal` is a subfield, then get all the FieldRefs that it
+          // refers to and then get all the values that alias with it.
+          forallRefersTo(dfsVal, [&](Node ref) {
+            // If this subfield refers to instance/mem results(input port), then
+            // add the output port FieldRefs that exist in the referenced module
+            // comb paths to the children.
+            handlePorts(ref, children);
+            // Get all the values that refer to this FieldRef, and add them to
```

Comment:
I guess this code path is very hot and we might want to avoid allocations in the inner-loop as much as possible. Can we precompute `aliasingValues` in preprocess? (not 100% confident that improves the execution time though).

Reply:
We could cache them, add another `DenseMap<Value, DenseSet<Value>>` , for a map of Value to all other values that alias. I will try this and check if it helps the runtime.